harness/determined/pytorch/_pytorch_trial.py

```diff
@@ -207,6 +383,103 @@ def _set_data_loaders(self) -> None:
                     )
                 self.validation_loader = validation_data
 
+    def _step_batch(self):
+        self.batches_trained += 1
+
+        # True epoch-based training is not supported. Epoch start/end is calculated with batch.
+        if self.context.is_epoch_start():
+            self.on_epoch_start(self.get_epoch_idx(self.batches_trained))
+
+        if self.context.is_epoch_end():
+            self.on_epoch_end(self.get_epoch_idx(self.batches_trained))
+
+        if issubclass(type(self.max_length), Batch):
+            self._train_step()
+
+    def _step_epoch(self):
+        self.epochs_trained += 1
+        if issubclass(type(self.max_length), Epoch):
+            self._train_step()
+
+    def _train_step(self):
+        self.steps_completed += 1
+        if self.steps_completed % self.validation_period == 0:
+            self.report_training_progress(self.searcher_op)
+            val_metrics = self._validate()
+            self.report_validation_metrics(self.steps_completed, val_metrics)
+            self.upload_tb_files()
+            self.stop_requested()
+        if self.steps_completed % self.checkpoint_period == 0:
+            self.report_training_progress(self.searcher_op)
+            self.checkpoint(already_exiting=False)
+            self.upload_tb_files()
+            self.stop_requested()
+
+    def stop_requested(self):
+        if self.core_context.distributed.rank == 0:
+            if self.core_context.preempt.should_preempt():
+                raise ShouldExit()
+        if self.context.get_stop_requested():
+            raise ShouldExit()
+
+    def _steps_remaining(self):
+        return self.max_length.value - self.steps_completed
+
+    def _train(self, train_unit, min_checkpoint_period, min_validation_period):
```

Comment:
maybe call this `_train_for_op()` or something?

Also, this might feel a little cleaner if you passed the training length or maybe the `op` into this function rather than setting it on `self`?  I don't feel strongly about that though... if you don't think it's worth it don't bother.